File: desing-pattens-1/src/modelo/implementacao/desconto/DescontoMaisDeCincoItens.java
package modelo.implementacao.desconto;

import interfaces.IDesconto;
import modelo.Orcamento;

public class DescontoMaisDeCincoItens implements IDesconto {

	IDesconto proximo;

	@Override
	public double calcula(Orcamento orcamento) {

		if (orcamento.getItens().size() > 5) {
			return 0.1;
		}
		return proximo.calcula(orcamento);
	}

	@Override
	public void setProximo(IDesconto desconto) {

		proximo = desconto;
	}
}

File: desing-pattens-1/src/main/java/modelo/implementacao/desconto/DescontoMaisDeCincoItens.kt
package main.java.modelo.implementacao.desconto

import main.java.interfaces.IDesconto
import main.java.modelo.Orcamento

class DescontoMaisDeCincoItens : IDesconto {

    internal var proximo: IDesconto

    override fun calcula(orcamento: Orcamento): Double {

        return if (orcamento.itens.size > 5) {
            0.1
        } else proximo.calcula(orcamento)
    }

    @override fun setProximo(desconto: IDesconto) {

        proximo = desconto
    }
}
